### ISSUE TITLE
cdk bootstrap - no destroy option

### ISSUE THREAD

**Issue by ygoodmn** [aws/aws-cdk#986]
Steps 

1. run 'cdk bootstrap' creates S3 Bootstrap Bucket via cloudformation template.

2. run 'cdk bootstrap destroy'  or 'cdk destroy'

Expected: 
Informs you you need to remove the bootstrap bucket manually and then delete the cloudformation. (perhaps also the name and region with links)
Or : Removes the cloudformation S3 bucket , which could get below error.
-----The bucket you tried to delete is not empty (Service: Amazon S3; Status Code: 409; Error Code: BucketNotEmpty; Request ID:
Or:  Option to delete the s3 bucket if empty

Current:  S3 bucket not removed or no option to run. 

Workaround:
Open console our use CLI to remove the s3 bucket created and the cloudformatin template

Note: Is it a standard use case to always run cdk bootstrap before initial deploy.

**Comment by RomainMuller** [comment]
The standard practice would be to run `cdk bootstrap` once per account/region pair you want to deploy CDK stuff into (using `cdk deploy`). You shouldn't need to destroy this ever (but I reckon you can if you don't intend on using the CDK anymore, or are not using the CDK toolkit for deployment).

**Comment by eladb** [comment]
I think there are use cases where people want to destroy the bootstrap stack, let alone cleaning the bucket up. Technically those buckets can fill up quite easily and incur unwanted costs for users.

I think we should support something like:

```console
$cdk bootstrap --destroy
```

And:

```console
$ cdk boostrap --clean
```

**Comment by jungseoklee** [comment]
> Is it a standard use case to always run `cdk bootstrap` before initial deploy.

As far as I know, if a template size is bigger than 50KB, `cdk bootstrap` turns into an essential step in order to upload the template to S3.

**Comment by Dzhuneyt** [comment]
Is it enough to just destroy the CloudFormation stack `CDKToolkit` that was created by the `cdk bootstrap` command as a measure to reverse the operation?

**Comment by Dzhuneyt** [comment]
This is also useful, for anyone who landed here: https://github.com/aws/aws-cdk/issues/1812#issuecomment-465983177

**Comment by shivlaks** [comment]
> `olkit` that was created by the `cdk bootstrap` command as a measure to reverse the operation

IIRC, If the bucket is non-empty, then the CloudFormation delete step will skip it. Only empty buckets can be deleted by destroying the stack.

**Comment by owenashurst** [comment]
Is it planned to add a delete command for this or are we just relying on the user deleting the CDKToolkit stack? Just out of curiosity

**Comment by nerdysolutions** [comment]
Just ran into this issue. i had a frontend that failed to deploy. part of that frontend deployment was a bunch of s3 buckets. i accidentally deleted the cdk staging bucket and got stuck not being able to do anything. would make total sense to be able to delete the bootstrap.

**Comment by nerdysolutions** [comment]
Thanks to this thread i was able to go into CloudFormation and delete the CDKToolkit template (otherwise I would have no idea) and then re-run cdk bootstrap after that. However, I agree it should be an option in the cli.

**Comment by moltar** [comment]
It is important to be able to un-bootstrap for the sakes of testing everything from 0 to 100%.

To clean up fast now:

```sh
aws cloudformation delete-stack --stack-name CDKToolkit
aws s3 ls # copy the name
aws s3 rb s3://cdktoolkit-stagingbucket-abcdef # replace the name here
```

**Comment by apoorvmote** [comment]
I have been trying out cdk and I am really bad at naming. I named everything `test`. Now when I read cloudformation template I can't tell what is meaning of each `test`. But I was able to try out everything. 

Now I would like to destory bootstrap s3 bucket and there is no option. Other than manual nuclear force delete.

@RomainMuller 
> The standard practice would be to run cdk bootstrap once per account/region pair you want to deploy CDK stuff into (using cdk deploy). You shouldn't need to destroy this ever (but I reckon you can if you don't intend on using the CDK anymore, or are not using the CDK toolkit for deployment).

Also within same account and within same region I plan to launch multiple project and I don't want to mix all of them into single cdk project. I want each project to have their own independant cdk project.

**Comment by apoorvmote** [comment]
# **PLEASE HELP**
I manually deleted cdktoolkit-stagingbucket-xxx and now I tried to create another project and deploy.

Its successfully bootstrapping but it fails to deploy with following error.

`[100%] fail: No bucket named 'cdktoolkit-stagingbucket-xxx'. Is account {account#} bootstrapped?`

And here is when I `cdk bootsrap` again.

```
 ⏳  Bootstrapping environment aws://{account#}/eu-west-1...
 ✅  Environment aws://{account#}/eu-west-1 bootstrapped (no changes).
```

Again when I deploy it fails with error above

**Comment by 0xVesion** [comment]
> It is important to be able to un-bootstrap for the sakes of testing everything from 0 to 100%.
> 
> To clean up fast now:
> 
> ```shell
> aws cloudformation delete-stack --stack-name CDKToolkit
> aws s3 ls | grep cdktoolkit # copy the name
> aws s3 rb s3://cdktoolkit-stagingbucket-abcdef # replace the name here
> ```
Thank you for this snippet! 👍 
I needed to add `--force` to the last command to allow the deletion of a non empty s3 bucket.

**Comment by moltar** [comment]
Thanks @0xVesion for the correction. I edited my original comment to add the `--force`

**Comment by hertzsprung** [comment]
I couldn't successfully delete the CDKToolkit stack until the staging bucket was emptied.  Building on earlier comments from @moltar and @0xVesion, I use:

````bash
cdk destroy
aws s3 rm --recursive s3://$(aws s3 ls | grep cdktoolkit | cut -d' ' -f3) # empty the cdktoolkit staging bucket
aws cloudformation delete-stack --stack-name CDKToolkit
````

**Comment by mimozell** [comment]
Would be nice to see some action on this item :)

**Comment by nikita-sheremet-clearscale** [comment]
> The standard practice would be to run `cdk bootstrap` once per account/region pair you want to deploy CDK stuff into (using `cdk deploy`). You shouldn't need to destroy this ever (but I reckon you can if you don't intend on using the CDK anymore, or are not using the CDK toolkit for deployment).

But what if CDK bootstrap deploys bucket and ECR repo? How will pay for content in s3 bucket even when it is not used? Who will pay for ECR repository? What if one of my cloud applications is needed to remove? I will run cdk destroy and still pay for all file/image in bucket and ECR repo? What if some application pushed sensitive content via assets - I have to manually process all bootstrap bucket content to find and clean assets for one application - this is a lot of work.

Also what about the security? Some with access to CdkToolKit stack can break everything in my region and account?

In my opinion, it is much better to have multiple bootstrapped stacks per several (but not all!) cloud applications.

**Comment by nikita-sheremet-clearscale** [comment]
Also, look for your competitors in serverless framework - they create a bootstrap bucket per deployment or allow you to set this bucket and create it in a separate stack. They do not force users to bootstrap once.

**Comment by alex-unikorny** [comment]
Waiting for this, the concept to recreate 0% to 100% all cloud resources is a must on an automated environment :), I know that the first resources like some initial IAM policies etc could not be automated at all but. All of resources that you can create with the initial IAM permissions should be deleted and created on easy way.

Any due date for that? Thanks in advance!

**Comment by TheRealAmazonKendra** [comment]
Related https://github.com/aws/aws-cdk/issues/19172

**Comment by evgenyka** [comment]
Removing the bootstrap bucket could result in failures in future deployments, so we need to proceed with caution.
We will think about this one as part of https://github.com/aws/aws-cdk-rfcs/issues/64.

**Comment by gshpychka** [comment]
> Removing the bootstrap bucket could result in failures in future deployments, so we need to proceed with caution. We will think about this one as part of [aws/aws-cdk-rfcs#64](https://github.com/aws/aws-cdk-rfcs/issues/64).

Sure, removing anything in the CDK toolkit stack could result in failures in future deployments - but people want to have that ability. The linked RFC does not cover destroying the `CDKToolkit` stack from the CLI, I believe that's a different issue.

**Comment by pahud** [comment]
@cwacek  Yes we are actively watching every issue and I agree with you. CDK CLI should allow user to easily "un-bootstrap" as well as clean up all resources created during and after the bootstrap process including the assets in the staging S3 buckets and ECR repo.

For me the experience should be like

$ cdk bootstrap --destroy
or
$ cdk bootstrap --clean

as mentioned initially by @eladb 

I am reopening this issue now and I'll bring up to the produce engineering immediately for review.

**Comment by gshpychka** [comment]
> I am currently in a situation where I need to rebootstrap (due to an update to the CDK library version), have deleted the CDKToolkit stack, but now cannot bootstrap because the asset bucket already exists. 

You don't need to destroy the stack to update it - simply run the bootstrap command and the stack will be updated to the latest version.

If CDK did have a command to destroy the bootstrap stack, it would have the same issue - the bucket would need to be deleted manually.

**Comment by rix0rrr** [comment]
We will put this in our backlog. @gshpychka is correct, you only need this for the cases where you intend to stop using CDK.

- We should print that and get user confirmation
- We should clean up resources (delete all S3 keys etc) and force deletion of RETAIN resources.

**Comment by cwacek** [comment]
Thanks for the response.

While I still think this is a valid issue, I did discover the reason I was
being forced to delete and re-create the bootstrap stack: an odd
interaction with the cdk-enterprise-iac
ConvertInlinePoliciesToManaged Aspect which changed a policy resource type
on cdk version update for some reason.
